data/dfinsupp.lean

```diff
@@ -0,0 +1,868 @@
+import data.finset data.set.finite algebra.big_operators algebra.module algebra.pi_instances
+
+universes u u₁ u₂ v v₁ v₂ v₃ w x y l
+
+variables (ι : Type u) (β : ι → Type v)
+
+instance decidable_zero_symm {γ : Type w} [has_zero γ] [decidable_pred (eq (0 : γ))] : decidable_pred (λ x, x = (0:γ)) :=
+λ x, decidable_of_iff (0 = x) eq_comm
+
+namespace dfinsupp
+
+@[reducible] protected def finset (s : finset ι) : Type* :=
+Π i : (↑s : set ι), β i.1
+
+def pre : Type* :=
+Σ s, dfinsupp.finset ι β s
+
+variable [Π i, has_zero (β i)]
+
+def rel (f g : pre ι β) : Prop :=
+(∀ i (h1 : i ∈ f.1) (h2 : i ∈ g.1), f.2 ⟨i, h1⟩ = g.2 ⟨i, h2⟩)
+∧ (∀ i (h1 : i ∈ f.1) (h2 : i ∉ g.1), f.2 ⟨i, h1⟩ = 0)
+∧ (∀ i (h1 : i ∉ f.1) (h2 : i ∈ g.1), g.2 ⟨i, h2⟩ = 0)
+
+instance : setoid (pre ι β) :=
+{ r := relation.trans_gen (rel ι β),
+  iseqv := ⟨λ f, relation.trans_gen.single
+      ⟨λ _ _ _, rfl, by split; intros; cc⟩,
+    λ f g H, relation.trans_gen.rec_on H
+      (λ g H, relation.trans_gen.single
+        ⟨λ _ _ _, (H.1 _ _ _).symm,
+        λ _ _ h2, H.2.2 _ h2 _,
+        λ _ h2 _, H.2.1 _ _ h2⟩)
+      (λ b c h1 h2 h3, relation.trans_gen.head
+        (⟨λ _ _ _, (h2.1 _ _ _).symm,
+          λ _ _ h4, h2.2.2 _ h4 _,
+          λ _ h4 _, h2.2.1 _ _ h4⟩) h3),
+    λ x y z, relation.trans_gen.trans⟩ }
+
+end dfinsupp
+
+def dfinsupp [Π i, has_zero (β i)] : Type* :=
+quotient (dfinsupp.setoid ι β)
+
+variables {ι β}
+
+infix ` →ₚ `:25 := dfinsupp

```

Comment:
this notation is a bit confusing. Can we have a binder notation like `Π₀` for this?